### PR TITLE
update MPFR to 4.2.0

### PR DESCRIPTION
**Pull request by simonbyrne**
Updates MPFR library, and makes use of new trig functions

### CODE REVIEW

**Comment by giordano** [comment]
Should add tests for the new methods?

**Comment by simonbyrne** [comment]
I think it should be covered by tests in `test/math.jl`, but will check

**Comment by simonbyrne** [comment]
Hmm, looks like not. I'll update them

**Comment by PallHaraldsson** [comment]
Possibly you want to define e.g. tanpi since it's already implemented there: "mpfr_tanpi, mpfr_acospi, mpfr_asinpi, mpfr_atanpi and mpfr_atan2pi"? It seems trivial to add, for MPFR (though not for Float64).

It's not yet in Base for floats, but [a]tabpi is in C2x standard (and IEEE 754:2019 and C#/dotnet seems to implemente it, issue 20137 there):
https://en.cppreference.com/w/c/experimental/fpext4

https://grouper.ieee.org/groups/msc/ANSI_IEEE-Std-754-2019/background/tanpi.txt
>754 and tanPi - exceptionally interesting
>
>754-2008 defined sin, cos, tan, asin, acos, atan, and
atan2, but only sinPi, cosPi, atanPi, and atan2Pi.    What
happened to tanPi, asinPi, acosPi? [..]

I see that you want to bring MPFR support in line with Base (floats), to be fully generic, so it's less than ideal to only have some functions for big/MPFR, not regular floats... the opposite problem, but we want to implement those at least eventually. We could actually do it by calling MPFR and then truncate to e.g. Float64? So not fast but could be improved later...

**Comment by simonbyrne** [comment]
I'm just adding methods to functions which are already are defined in Base, and which have corresponding functions that were added to MPFR in this release.

I think adding the additional functions to Base is a good idea, but should be a separate PR

**Comment by simonbyrne** [comment]
Not sure if we want to backport or not?

**Comment by oscardssmith** [comment]
why would we backport? This looks to me a lot like a new feature.

File: base/mpfr.jl [hunk 1]
```diff
@@ -791,13 +792,40 @@ for f in (:sin, :cos, :tan, :sec, :csc, :acos, :asin, :atan, :acosh, :asinh, :at
...
+for f in (:sin, :cos, :tan)
+    @eval begin
+        function ($(Symbol(f,:d)))(x::BigFloat)
+            isnan(x) && return x
+            z = BigFloat()
+            ccall(($(string(:mpfr_,f,:u)), :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Culong, MPFRRoundingMode), z, x, 360, ROUNDING_MODE[])
+            isnan(z) && throw(DomainError(x, "NaN result for non-NaN input."))
+            return z
+        end
+        function ($(Symbol(:a,f,:d)))(x::BigFloat)
```
Comment (giordano) [inline base/mpfr.jl:813]: Does this define an `atand` method which is overridden below?

**Comment by StefanKarpinski** [comment]
Updating the library could be backported but adding methods should not.

**Comment by simonbyrne** [comment]
It doesn't look like there are any important bug fixes: https://www.mpfr.org/mpfr-4.2.0/#changes (their printing fix actually broke our printing, hence the change in this PR)

File: base/mpfr.jl (hunk 1 above)
Comment (simonbyrne) [reply base/mpfr.jl:813]: no, this is the 1-arg `atand`, below is the 2-arg version.

File: test/math.jl [hunk 2]
```diff
@@ -512,8 +517,8 @@ end
                 @test my_eq(sincospi(one(T)/convert(T,6))[1], 0.5)
                 @test_throws DomainError sind(convert(T,Inf))
                 @test_throws DomainError cosd(convert(T,Inf))
-                T != Float32 && @test my_eq(cospi(one(T)/convert(T,3)), 0.5)
-                T != Float32 && @test my_eq(sincospi(one(T)/convert(T,3))[2], 0.5)
+                fT == Float64 && @test my_eq(cospi(one(T)/convert(T,3)), 0.5)
+                fT == Float64 && @test my_eq(sincospi(one(T)/convert(T,3))[2], 0.5)
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply test/math.jl:521]: It isn't, if `T == Rational{Int}`, then `fT == Float64`

[the remark this answers is not in the document]
Comment (simonbyrne) [reply test/math.jl:521]: Basically, this test only holds for `fT = Float64` (and probably should be constructed differently)

**Review by giordano — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Wow, all green CI! Haven't seen that recently.

**Pull request merged**